lib/schedulers.go

```diff
@@ -0,0 +1,331 @@
+			errors = append(errors,
+				fmt.Errorf("scheduler %s has errors: %s", name, concatErrors(schedErr, ", ")))
+		}
+	}
+	return errors
+}
+
+// GetSortedSchedulerConfigs returns a slice with the scheduler configurations,
+// sorted in a consistent and predictable manner. It is useful when we want or
+// have to avoid using maps with string keys (and tons of string lookups in
+// them) and avoid the unpredictable iterations over Go maps. Slices allow us
+// constant-time lookups and ordered iterations.
+//
+// The configs in the returned slice will be sorted by their start times in an
+// ascending order, and alphabetically by their names (which are unique) if
+// there are ties.
+func (scs SchedulerConfigMap) GetSortedSchedulerConfigs() []SchedulerConfig {
+	configs := make([]SchedulerConfig, len(scs))
+
+	// Populate the configs slice with sorted scheduler configs
+	i := 0
+	for _, config := range scs {
+		configs[i] = config // populate the slice in an unordered manner
+		i++
+	}
+	sort.Slice(configs, func(a, b int) bool { // sort by (start time, name)
+		switch {
+		case configs[a].GetStartTime() < configs[b].GetStartTime():
+			return true
+		case configs[a].GetStartTime() == configs[b].GetStartTime():
+			return strings.Compare(configs[a].GetName(), configs[b].GetName()) < 0
+		default:
+			return false
+		}
+	})
+
+	return configs
+}
+
+// GetFullExecutionRequirements combines the execution requirements from all of
+// the configured schedulers. It takes into account their start times and their
+// individual VU requirements and calculates the total VU requirements for each
+// moment in the test execution.
+func (scs SchedulerConfigMap) GetFullExecutionRequirements(executionSegment *ExecutionSegment) []ExecutionStep {
+	sortedConfigs := scs.GetSortedSchedulerConfigs()
+
+	// Combine the steps and requirements from all different schedulers, and
+	// sort them by their time offset, counting the schedulers' startTimes as
+	// well.
+	type trackedStep struct {
+		ExecutionStep
+		configID int
+	}
+	trackedSteps := []trackedStep{}
+	for configID, config := range sortedConfigs { // orderly iteration over a slice
+		configStartTime := config.GetStartTime()
+		configSteps := config.GetExecutionRequirements(executionSegment)
+		for _, cs := range configSteps {
+			cs.TimeOffset += configStartTime // add the scheduler start time to the step time offset
+			trackedSteps = append(trackedSteps, trackedStep{cs, configID})
+		}
+	}
+	// Sort by (time offset, config id). It's important that we use stable
+	// sorting algorithm, since there are could be steps with the same time from
+	// the same scheduler and their order is important.
+	sort.SliceStable(trackedSteps, func(a, b int) bool {
+		switch {
+		case trackedSteps[a].TimeOffset < trackedSteps[b].TimeOffset:
+			return true
+		case trackedSteps[a].TimeOffset == trackedSteps[b].TimeOffset:
+			return trackedSteps[a].configID < trackedSteps[b].configID
+		default:
+			return false
+		}
+	})
```

Comment:
I would prefer if everything upto this point is a different method ... something like `SortedSteps` so we can have the whole function fit on a single page ... To this end I more and more dislike the super long sections of comments in the middle of the functions. 
Some are okay(all of them are useful IMO) but at some point a 10 line function is 80 lines of comments in between all the actual code lines. I think that in most cases you can put the comment on top.

Reply:
>I would prefer if everything upto this point is a different method ... something like SortedSteps so we can have the whole function fit on a single page

hmm I'd argue that this won't be better of in a separate function - splitting it up doesn't make sense for a few reasons:
- the code up to here doesn't make logical sense on it's own - it won't be usable for anything else besides the rest of this method
- the minor `trackedStep` type is used in both parts of the method, and nowhere else
- the whole method is only 60 lines of code (without the comments), hardly a monster... and a lot of those lines is actually due to the verbosity of Go - I to implement a `sum()` function ffs...

> To this end I more and more dislike the super long sections of comments in the middle of the functions.

Sometimes I do as well, but in this specific function, as you've seen, the comment below serves the role of visually splitting up the 2 main sections of the function body, besides explaining what the second section actually does.

>  I think that in most cases you can put the comment on top.

My rule of thumb is for the function/method comments to describe what the function/method does, while the inline comments in its body to describe how it does it. In this case, if I push the inline comments to the top, they won't be useful for users of the function and the function code will be more difficult to understand for future readers.

Reply:
ping @MStoykov  (and @imiric , @cuonglm ) :arrow_up: is my logic sound, or do you still want me to split up that very well commented method? :smile:

Reply:
`GetFullExecutionRequirements` LGTM as is. It's not overly long, and the inline comments are relevant and helpful. :+1:

OTOH, some of the executor methods like `Run` would be better off split, but it's, umm, tricky, to say the least. :smile:

Reply:
Yeah, tricky is one word for it :sweat_smile: Though it can be done, case in point how I recently split the `externally-controlled` executor's `Run()` method: https://github.com/loadimpact/k6/pull/1007/commits/2ebe3dfdd6c3ef97decbd9ac80c6747dedb3aac1 

Whether the added complexity of that approach is worth it is another topic... For that specific method - for sure yes. For the other simpler ones - maybe?

Reply:
> how ... probably not important in a lot of the cases ;)

Well, it's very important in this particular case :wink: And if that helper function would have been useful anywhere else, I would have extracted it, but it isn't... I'm resolving this, since this function isn't even all that long - it's just very full of comments... Which would be useful when someone  has to refactor this in the future, since these comments explain important implementation details that can't be encoded with the type system...